libs/traits/src/lib.rs

```diff
@@ -555,6 +583,78 @@ pub trait TokenSwaps<Account> {
 	) -> Result<Self::BalanceIn, DispatchError>;
 }
 
+/// A representation of a currency swap in process.
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, MaxEncodedLen)]
+pub struct SwapState<AmountIn, AmountOut, Currency> {
+	/// Swap not yet processed with the pending outcomming amount
+	pub remaining: Swap<AmountOut, Currency>,
+
+	/// Amount of incoming currency already swapped
+	pub swapped_in: AmountIn,
+
+	/// Amount of incoming currency already swapped denominated in outgoing
+	/// currency
+	pub swapped_out: AmountOut,
+}
+
+/// Used as result of `Pallet::apply_swap()`
+/// Amounts are donominated referenced by the `new_swap` paramenter given to
+/// `apply_swap()`
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct SwapStatus<Amount> {
+	/// The incoming amount already swapped and available to use.
+	pub swapped: Amount,
+
+	/// The outgoing amount pending to be swapped
+	pub pending: Amount,
+}
+
+/// Trait to perform swaps without handling directly an order book
+pub trait Swaps<AccountId> {
+	type Amount;
+	type CurrencyId;
+	type SwapId;
+
+	/// Apply a swap over a current possible swap state.
+	/// - If there was no previous swap, it adds it.
+	/// - If there was a swap in the same direction, it increments it.
+	/// - If there was a swap in the opposite direction:
+	///   - If the amount is smaller, it decrements it.
+	///   - If the amount is the same, it removes the inverse swap.
+	///   - If the amount is greater, it removes the inverse swap and create
+	///     another with the excess
+	///
+	/// The returned status contains the swapped amount after this call
+	/// (denominated in the incoming currency) and the pending amounts to be
+	/// swapped.
+	fn apply_swap(
+		who: &AccountId,
+		swap_id: Self::SwapId,
+		swap: Swap<Self::Amount, Self::CurrencyId>,
+	) -> Result<SwapStatus<Self::Amount>, DispatchError>;
+
+	/// Returns the pending amount for a pending swap. The direction of the swap
+	/// is determined by the `from_currency` parameter. The amount returned is
+	/// denominated in the same currency as the given `from_currency`.
+	fn pending_amount(
+		who: &AccountId,
+		swap_id: Self::SwapId,
+		from_currency: Self::CurrencyId,
+	) -> Result<Self::Amount, DispatchError>;
+
+	/// Check that validates that if swapping pair is supported.
+	fn valid_pair(currency_in: Self::CurrencyId, currency_out: Self::CurrencyId) -> bool;
+
+	/// Makes a conversion between 2 currencies using the market ratio between
+	/// them
+	// TODO: Should be removed after #1723
+	fn convert_by_market(
+		currency_in: Self::CurrencyId,
+		currency_out: Self::CurrencyId,
+		amount_out: Self::Amount,
+	) -> Result<Self::Amount, DispatchError>;
```

Comment:
This should not be in the trait. Once https://github.com/centrifuge/centrifuge-chain/pull/1723 is fixed. Getting the market rate is very low level for this trait, I guess.